core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala

```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
+                      partitionAssigned(
+                        producer,
+                        tp.partition(),
+                        endOffsets.get(tp) - 1,
+                        assignedPartitions,
+                        dispatcher
+                      )
+                    )
+                }
+              }
+            }
+          )
+        )
+      )
+    } yield ()
+
+    def partitionRevoked(p: Partition, assignedPartitions: Ref[IO, Map[Partition, AssignedPartition]]): IO[Unit] =
+      for {
+        partitions <- assignedPartitions.get
+        _ <- partitions.get(p).traverse(_.shutdown)
+        _ <- assignedPartitions.update(_ - p)
+      } yield ()
+
+    def partitionAssigned(
+        producer: KafkaProducer[Array[Byte], Array[Byte]],
+        p: Partition,
+        endOffset: Offset,
+        assignedPartitions: Ref[IO, Map[Partition, AssignedPartition]],
+        dispatcher: Dispatcher[IO]
+    ): IO[Unit] = for {
+      (redeliverActor, shutdown) <- RedeliverActor.create(p, producer, config, helpers.clients).allocated
+      _ <- redeliverActor.tell(DoRedeliver)
+      _ <- assignedPartitions.update(
+        _.updated(
+          p,
+          AssignedPartition(new MarkersQueue(endOffset - 1), redeliverActor, shutdown, config, None, None, dispatcher)
+        )
+      )
+    } yield ()
+
+    def setupOffsetCommitting = for {
+      commitMarkerOffsetsActor <- CommitMarkerOffsetsActor
+        .create(
+          config.getMarkerTopic,
+          config.getMarkerConsumerOffsetGroupId,
+          helpers
+        )
+      _ <- Resource.eval(commitMarkerOffsetsActor.tell(DoCommit))
+    } yield commitMarkerOffsetsActor
+
+    for {
+      dispatcher <- Dispatcher.sequential[IO]
+      assignedPartitions <- Resource.eval(IO.ref(Map.empty[Partition, AssignedPartition]))
+      mailbox <- Resource.eval(Queue.unbounded[IO, ConsumeMarkersActorMessage])
+      markerConsumer <- helpers.createConsumer(
+        config.getMarkerConsumerGroupId,
+        classOf[MarkerKey.MarkerKeyDeserializer],
+        classOf[MarkerValue.MarkerValueDeserializer]
+      )
+      _ <- setupMarkerConsumer(markerConsumer, assignedPartitions, dispatcher)
+      commitMarkerOffsetsActor <- setupOffsetCommitting
+      _ <- stream(assignedPartitions, mailbox, commitMarkerOffsetsActor, markerConsumer).compile.drain.background
+      _ <- Resource.eval(logger.info("Consume markers actor setup complete"))
+    } yield new ConsumeMarkersActor(mailbox)
   }
 
-  override def receive: Receive = {
-    case DoConsume =>
-      try {
-        val markers = markerConsumer.poll(OneSecond).asScala
-        val now = System.currentTimeMillis()
-
-        markers.groupBy(_.partition()).foreach { case (partition, records) =>
-          assignedPartitions.get(partition) match {
-            case None =>
-              throw new IllegalStateException(s"Got marker for partition $partition: ${records.map(_.key())}, but partition is not assigned!")
```

Comment:
I wonder if this will work correctly in case of a rebalance - it's possible that we get some messages, rebalance happens, we remove the partitions from `assignedPartitions`, and only then handle the messages - which will end with an error

maybe enqueueing the partition added/revoked as actor messages would help? (we'd also be able to get rid of the `Ref`) But then, we can receive messages for which we haven't yet added partitions ...

Reply:
maybe a solution would be to handle parition-revoked as an actor message (which would be handled only *after* the current receive finishes), and instead of handling partition-assigned here, moved it to the consuming logic - when we encounter a marker for an unknown partition, simply create an appropriate consumer (that is - we'd create `AssignedPartition` lazily)

Reply:
What would happen if we just ignore this state (for example just log a warning)?

Reply:
hm well couldn't it produce a false warning? my concern here would be that we could get a race condition, where the partition is added (w/ the callback being invoked), and at the same time new messages start arriving. So if this happened, not only we would get a false warning, but we would also loose a message. But  maybe messages aren't delivered until all the callback complete? That would require verification on the kafka side